subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+	    handle->description->type == AMOD_TYPE_OUTPUT ||
+	    handle_disconnect->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle_disconnect->description->type == AMOD_TYPE_INPUT) {
+		LOG_DBG("Disonnection of modules %s from %s, is not supported",
+			handle_disconnect->name, handle->name);
+		return -ENOTSUP;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED ||
+	    handle_disconnect->state == AMOD_STATE_UNDEFINED) {
+		LOG_DBG("A module is an invalid state or type for disconnection");
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&handle->dest_mutex, K_FOREVER);
+
+	if (handle == handle_disconnect) {
+		handle->data_tx = false;
+		handle->dest_count -= 1;
+	} else {
+		if (!sys_slist_find_and_remove(&handle->hdl_dest_list, &handle_disconnect->node)) {
+			LOG_DBG("Connection to module %s has not been found for module %s",
+				handle_disconnect->name, handle->name);
+			return -EALREADY;
+		} else {
+			LOG_DBG("Disconnect module %s from module %s", handle_disconnect->name,
+				handle->name);
+			handle->dest_count -= 1;
+		}
+	}
+
+	k_mutex_unlock(&handle->dest_mutex);
+
+	return 0;
+}
+
+/**
+ * @brief Start processing data in the module given by handle.
+ */
+int audio_module_start(struct audio_module_handle *handle)
+{
+	int ret;
+
+	if (handle == NULL) {
+		LOG_DBG("Module handle is NULL");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED) {
+		LOG_DBG("Module %s in an invalid state, %d, for start", handle->name,
+			handle->state);
+		return -ENOTSUP;
+	}
+
+	if (handle->state == AMOD_STATE_RUNNING) {
+		LOG_DBG("Module %s already running", handle->name);
+		return -EALREADY;
+	}
+
+	if (handle->description->functions->start != NULL) {
+		ret = handle->description->functions->start(
+			(struct audio_module_handle_private *)handle);
+		if (ret < 0) {
+			LOG_DBG("Failed user start for module %s, ret %d", handle->name, ret);
+			return ret;
+		}
+	}
+
+	handle->previous_state = handle->state;
+	handle->state = AMOD_STATE_RUNNING;
+
+	return 0;
+}
+
+/**
+ * @brief Stop processing data in the module given by handle.
+ */
+int audio_module_stop(struct audio_module_handle *handle)
+{
+	int ret;
+
+	if (handle == NULL) {
+		LOG_DBG("Module handle is NULL");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_STOPPED) {
+		LOG_DBG("Module %s already stopped", handle->name);
+		return -EALREADY;
+	}
+
+	if (handle->state != AMOD_STATE_RUNNING) {
+		LOG_DBG("Module %s in an invalid state, %d, for stop", handle->name, handle->state);
+		return -ENOTSUP;
+	}
+
+	if (handle->description->functions->stop != NULL) {
+		ret = handle->description->functions->stop(
+			(struct audio_module_handle_private *)handle);
+		if (ret < 0) {
+			LOG_DBG("Failed user pause for module %s, ret %d", handle->name, ret);
+			return ret;
+		}
+	}
+
+	handle->previous_state = handle->state;
+	handle->state = AMOD_STATE_STOPPED;
+
+	return 0;
+}
+
+/**
+ * @brief Send a data buffer to a module, all data is consumed by the module.
```

Comment:
If we send data to a module, would not that be `audio_module_data_rx` ?

Reply:
I think we need to discuss the RX/TX naming.